### PR TITLE
`azurerm_role_assignment`: Fix assignments to resources

### LINKED ISSUE #12057: Role assignment issue introduced in 2.62 pinning to 2.61 resolves issue
Following was introduced in 2.62 pinning provider to 2.6.1 resolved.

[0mWhen applying changes to
module.Sitepod_AKS_Cluster.azurerm_role_assignment.route_table[0], provider
"registry.terraform.io/hashicorp/azurerm" produced an unexpected new value:
Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
[0m[0m
[31m
[1m[31mError: [0m[0m[1mProvider produced inconsistent result after apply[0m

[0mWhen applying changes to
module.Sitepod_AKS_Cluster.azurerm_role_assignment.virtual_network[0],
provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new
value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
[0m[0m

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

terraform 14.5

[0m[1mInitializing provider plugins...[0m
- Finding latest version of hashicorp/random...
- Finding hashicorp/vault versions matching "~> 2.19.0"...
- Finding hashicorp/azurerm versions matching ">= 2.56.0"...
- Finding nbering/ansible versions matching ">= 1.0.4"...
- Finding latest version of hashicorp/null...
- Finding dimensiondataresearch/ddcloud versions matching "3.0.6"...
- Installing hashicorp/azurerm v2.62.0...
- Installed hashicorp/azurerm v2.62.0 (self-signed, key ID [0m[1m34365D9472D7468F[0m[0m)
- Installing nbering/ansible v1.0.4...
- Installed nbering/ansible v1.0.4 (self-signed, key ID [0m[1m1A9FBF0FFD10BE33[0m[0m)
- Installing hashicorp/null v3.1.0...
- Installed hashicorp/null v3.1.0 (self-signed, key ID [0m[1m34365D9472D7468F[0m[0m)
- Installing dimensiondataresearch/ddcloud v3.0.6...
- Installed dimensiondataresearch/ddcloud v3.0.6 (self-signed, key ID [0m[1m89A0920B45513186[0m[0m)
- Installing hashicorp/random v3.1.0...
- Installed hashicorp/random v3.1.0 (self-signed, key ID [0m[1m34365D9472D7468F[0m[0m)
- Installing hashicorp/vault v2.19.1...
- Installed hashicorp/vault v2.19.1 (self-signed, key ID [0m[1m34365D9472D7468F[0m[0m)

### Terraform Configuration Files

resource "azurerm_role_assignment" "route_table" {
  count = var.aks_cluster_identity_type == "UserAssigned" ? 1 : 0

  scope                = data.azurerm_subnet.aks_cluster.route_table_id
  role_definition_name = "Network Contributor"
  principal_id         = azurerm_user_assigned_identity.aks_cluster[0].principal_id
}
resource "azurerm_role_assignment" "virtual_network" {
  count = var.aks_cluster_identity_type == "UserAssigned" ? 1 : 0

  scope                = data.azurerm_virtual_network.aks_cluster.id
  role_definition_name = "Network Contributor"
  principal_id         = azurerm_user_assigned_identity.aks_cluster[0].principal_id
}

### LINKED ISSUE #12060: Error in azurerm_role_assignment
In this code
```
resource "azurerm_role_assignment" "role_assignment_name" {
  scope                = var.rbac_scope_id
  role_definition_name = var.rbac_role_name
  principal_id         = var.principal_id
}
```
I'm getting this error
```
2021-06-04T09:34:10.3860839Z Error: Provider produced inconsistent result after apply
2021-06-04T09:34:10.3861369Z 
2021-06-04T09:34:10.3861847Z When applying changes to
2021-06-04T09:34:10.3862760Z module.provision_keyvaults[0].module.kv_create["pwd-002"].module.kv_roles_aadgroups["Contributor"].module.rbac-assign-aadgroup-by-role-name.azurerm_role_assignment.role_assignment_name,
2021-06-04T09:34:10.3863640Z provider "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced
2021-06-04T09:34:10.3864333Z an unexpected new value: Root resource was present, but now absent.
2021-06-04T09:34:10.3864899Z 
2021-06-04T09:34:10.3865489Z This is a bug in the provider, which should be reported in the provider's
2021-06-04T09:34:10.3866035Z own issue tracker.
```
providers.tf
```
provider "azurerm" {
  alias           = "main"
  tenant_id       = var.tenant_id
  subscription_id = var.providers_azurerm["main"].subscription_id

  features {
    key_vault {
      purge_soft_delete_on_destroy = true
      recover_soft_deleted_key_vaults = true
    }
  }
}

provider "azurerm" {
  tenant_id       = var.tenant_id
  subscription_id = var.providers_azurerm["subscription"].subscription_id
  
  features {
    key_vault {
      purge_soft_delete_on_destroy = true
      recover_soft_deleted_key_vaults = true
    }
  }
}
```

If more files would be needed, just ask
[RawTF.TXT](https://github.com/terraform-providers/terraform-provider-azurerm/files/6597481/RawTF.TXT)

### LINKED ISSUE #12074: Seems like azurerm_role_assignment is trying to get state from wrong Azure Uri in AzureRm version 2.62
### Terraform Version

```
Terraform v0.15.5
on windows_amd64
+ provider registry.terraform.io/hashicorp/azuread v1.5.0
+ provider registry.terraform.io/hashicorp/azurerm v2.62.0
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
```

### Terraform Configuration Files

```terraform

```

### Debug Output
```TXT
2021-06-04T22:36:03.135+0200 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5.exe: [DEBUG] AzureRM Response for https://management.azure.com//subscriptions/<SUBSCRIPTION_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.Authorization/roleAssignments/<ROLE_ASSIGNMENT_ID>?api-version=2020-04-01-preview: 
2021-06-04T22:36:03.135+0200 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5.exe: HTTP/2.0 404 Not Found
...
2021-06-04T22:36:03.135+0200 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5.exe: {"error":{"code":"RoleAssignmentNotFound","message":"The role assignment '<ROLE_ASSIGNMENT_ID>' is not found."}}
2021-06-04T22:36:03.135+0200 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5.exe: [DEBUG] Role Assignment ID "/subscriptions/<SUBSCRIPTION_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.Storage/storageAccounts/<STORAGE_ACCOUNT_NAME>/providers/Microsoft.Authorization/roleAssignments/<ROLE_ASSIGNMENT_ID>" was not found - removing from state
2021-06-04T22:36:03.135+0200 [WARN]  Provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value for azurerm_role_assignment.example during refresh.
      - Root resource was present, but now absent
```

> NOTE: See the difference in the Uri Terraform uses vs. the actual Resource ID it says it checked. They do not match. 
>
> This is missing in the TF Get request: `providers/Microsoft.Storage/storageAccounts/<STORAGE_ACCOUNT_NAME>`

### Crash Output

N/A

### Expected Behavior
I would expect a re-run of `terraform apply` to work.

### Actual Behavior

Running `terraform apply` fails, since it says the `azurerm_role_assignment` already exists. This could indicate that the state isn't recorded.

### Steps to Reproduce

* Create a `azurerm_role_assignment` with the scope of `azurerm_storage_account`
* Run `terraform apply`, and the Role Assignmant is added
* Run `terraform apply` again, and it fails, saying it already exists

### Additional Context

Looking at the Trace Log, I notice that it calls an endpoint with a different ID than the Resource ID it says it has. Running `az rest -u <RESOURCE ID>` with the "correct" id works (see below). But the one TF calls, gives a 404, Not Found.

_Does **not** work (The URI Terraform uses)_:
```SH
az rest -u https://management.azure.com//subscriptions/<SUBSCRIPTION_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.Authorization/roleAssignments/<ROLE_ASSIGNMENT_ID>?api-version=2020-04-01-preview
```

**Works**:
```SH
az rest -u https://management.azure.com//subscriptions/<SUBSCRIPTION_ID>/resourceGroups/<RG_NAME>/providers/Microsoft.Storage/storageAccounts/<STORAGE_ACCOUNT_NAME>/providers/Microsoft.Authorization/roleAssignments/<ROLE_ASSIGNMENT_ID>?api-version=2020-04-01-preview
```

### References

The Issue was first reported in the Terraform repo:
https://github.com/hashicorp/terraform/issues/28886

### LINKED ISSUE #12078: Error: Provider produced inconsistent result after apply azurem
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

* `azurerm_role_assignment`

```
Terraform v0.15.5
on darwin_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.62.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
```

### Terraform Configuration Files

```hcl
resource "azurerm_role_assignment" "managed_identity_storage_assignment" {
  name               = uuid()
  scope              = var.storage_account_id
  role_definition_name = "Storage Blob Data Contributor"
  principal_id       = var.webapp_identity
}
```

### Panic Output

none

### Expected Behaviour

works

### Actual Behaviour

```
Error: Provider produced inconsistent result after apply
When applying changes to module.storage_account_permissions.azurerm_role_assignment.managed_identity_storage_assignment, provider
"provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

### Steps to Reproduce

Running plan from local or Azure DevOps workspace.

### LINKED ISSUE #12079: Cannot import azurerm_role_assignment for scope other than subscription or resource group
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.15.5
on darwin_amd64
+ provider registry.terraform.io/hashicorp/azuread v1.5.0
+ provider registry.terraform.io/hashicorp/azurerm v2.62.0
+ provider registry.terraform.io/hashicorp/local v2.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
```

### Affected Resource(s)

* `azurerm_role_assignment`

### Terraform Configuration Files

```hcl
module "dns_service_principal" {
  source   = "../../shared/terraform/modules/service-principal"
  name     = "playground-dns-challenge-sp"
}

data "azurerm_dns_zone" "dnszone" {
  name                = "company.net"
  resource_group_name = "myresourcegroup"
}

resource "azurerm_role_assignment" "dns_zone_contributor" {
  scope                = data.azurerm_dns_zone.dnszone.id
  role_definition_name = "DNS Zone Contributor"
  principal_id         = module.dns_service_principal.object_id
}
```

### Description

When upgrading from terraform 13.5 -> 14.5 (and provider 2.41.0 -> 2.62.0) role assignments that are scoped to specific resources that are not a resource group or subscription are no longer found and are removed from state. 

When trying to import the role assignments, I always get:

```
 Error: Cannot import non-existent remote object
│ While attempting to import an existing object to "azurerm_role_assignment.dns_zone_contributor", the provider detected that no object exists with the given
│ id. Only pre-existing objects can be imported; check that the id is correct and that it is associated with the provider's configured region or endpoint, or
│ use "terraform apply" to create a new remote object for this resource.
```

### Expected Behaviour

* Ideally resource doesn't need to be re-imported
* If it does need to be, it should be successfully

### Actual Behaviour

* Needs to be imported (Running plan shows this needs to be created, running apply fails as it already exists)
* Import fails

### Steps to Reproduce

1. `az role assignment list --scope /subscriptions/XXX/resourceGroups/myresourcegroup/providers/Microsoft.Network/dnszones/company.net` find the correct id from there and us in next 
2. `terraform import azurerm_role_assignment.dns_zone_contributor /subscriptions/XXX/resourceGroups/myresourcegroup/providers/Microsoft.Network/dnszones/company.net/providers/Microsoft.Authorization/roleAssignments/XXX`
3. Observe error above

### Important Factoids

From the debug logs, it looks like it's replacing the scope with the resource group scope:

```
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: [DEBUG] AzureRM Response for https://management.azure.com//subscriptions/XXX/resourceGroups/myresourcegroup/providers/Microsoft.Authorization/roleAssignments/XXX?api-version=2020-04-01-preview:
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: HTTP/2.0 404 Not Found
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Content-Length: 128
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Cache-Control: no-cache
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Content-Type: application/json; charset=utf-8
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Date: Sun, 06 Jun 2021 20:24:08 GMT
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Expires: -1
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Pragma: no-cache
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Set-Cookie: x-ms-gateway-slice=Production; path=/; secure; samesite=none; httponly
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: Strict-Transport-Security: max-age=31536000; includeSubDomains
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: X-Content-Type-Options: nosniff
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: X-Ms-Correlation-Request-Id: 2d624d05-0b33-7a32-427d-eff5a05be3be
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: X-Ms-Ratelimit-Remaining-Subscription-Reads: 11995
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: X-Ms-Request-Id: a1a5b7fc-e738-43ce-97ae-22b4063a9881
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: X-Ms-Routing-Request-Id: UKSOUTH:20210606T202409Z:XXX
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5:
2021-06-06T21:24:09.485+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: {"error":{"code":"RoleAssignmentNotFound","message":"The role assignment 'XXX' is not found."}}
2021-06-06T21:24:09.486+0100 [DEBUG] provider.terraform-provider-azurerm_v2.62.0_x5: [DEBUG] Role Assignment ID "/subscriptions/XXX/resourceGroups/myresourcegroup/providers/Microsoft.Network/dnszones/company.net/providers/Microsoft.Authorization/roleAssignments/XXX" was not found - removing from state
2021-06-06T21:24:09.486+0100 [WARN]  Provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value for azurerm_role_assignment.dns_zone_contributor during refresh.
      - Root resource was present, but now absent
```

### LINKED ISSUE #12087: Provider produced inconsistent result after apply (on role binding)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
Terraform v0.15.5
on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.62.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.3.1
+ provider registry.terraform.io/hashicorp/random v3.1.0

### Affected Resource(s)

* `azurerm_2.62.0`

### Terraform Configuration Files

```hcl
resource "azurerm_role_assignment" "azure_kubernetes_service_rbac_cluster_admin" {
  principal_id         = var.aad_group_objectid
  role_definition_name = var.aad_rbac_cluster_admin
  scope                = azurerm_kubernetes_cluster.k8s.id
}
```

### Debug Output

```
╷
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to azurerm_role_assignment.azure_kubernetes_service_rbac_cluster_admin, provider
│ "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an unexpected new value: Root resource was present, but now absent.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```

### Expected Behaviour

Creation of role assignment with success feddback in terraform

### Actual Behaviour

role assignment is indeed created, but Terraform repsonds with an error

### Steps to Reproduce

I'm not sure if this is liked to the creation of other resources as well. So I assume it is just about creating the role-binding-resource.

1. `terraform apply`

### Important Factoids

plain Azure public cloud

### PR DESCRIPTION
**Pull request by aristosvo**
Fixes  #12074 
Fixes  #12060 
Fixes  #12057
Fixes #12079
Fixes #12078
Fixes #12087

Related to/similar for [go-azure-helpers](https://github.com/hashicorp/go-azure-helpers): [this PR](https://github.com/hashicorp/go-azure-helpers/pull/79)

Done:
- [x] Reproduce the problem :)
```
❯ make acctests SERVICE='authorization' TESTARGS='-run=TestAccRoleAssignment_resourceScoped'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/authorization -run=TestAccRoleAssignment_resourceScoped -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/06/05 16:11:49 [DEBUG] not using binary driver name, it's no longer needed
2021/06/05 16:11:50 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccRoleAssignment_resourceScoped
=== PAUSE TestAccRoleAssignment_resourceScoped
=== CONT  TestAccRoleAssignment_resourceScoped
    testing.go:620: Step 1/2 error: Error running apply: exit status 1
        
        Error: Provider produced inconsistent result after apply
        
        When applying changes to azurerm_role_assignment.test, provider
        "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an
        unexpected new value: Root resource was present, but now absent.
        
        This is a bug in the provider, which should be reported in the provider's own
        issue tracker.
--- FAIL: TestAccRoleAssignment_resourceScoped (136.30s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/authorization 139.640s
FAIL
make: *** [acctests] Error 1
```
- [x] Diagnose the problem:
  - Previously the ID was saved raw and without transformation/checks. Because the roleAssignmentID parser doesn't contain a resource/scope part, it wasn't saved and therefore not (able to be) regenerated correctly based on saved parameters: `/subscriptions/<subscription>/resourceGroups/<resource_group>/providers/Microsoft.Storage/storageAccounts/<storage_account_name>/providers/Microsoft.Authorization/roleAssignments/<UUID>` became `/subscriptions/<subscription>/resourceGroups/<resource_group>/providers/Microsoft.Authorization/roleAssignments/<UUID>`

- [x] Implement the fix
```
❯ make acctests SERVICE='authorization' TESTARGS='-run=TestAccRoleAssignment_resourceScoped'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/authorization -run=TestAccRoleAssignment_resourceScoped -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/06/06 08:41:44 [DEBUG] not using binary driver name, it's no longer needed
2021/06/06 08:41:45 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccRoleAssignment_resourceScoped
=== PAUSE TestAccRoleAssignment_resourceScoped
=== CONT  TestAccRoleAssignment_resourceScoped
--- PASS: TestAccRoleAssignment_resourceScoped (191.83s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/authorization       195.451s
```

- [x] Clean up
  - Fix the unit tests of other resources by providing a general `id.SecondaryProvider`, which solved the problems for the resources which have multiple provider entries in their IDs

### CODE REVIEW

**Comment by njuCZ** [comment]
@aristosvo thanks for this PR. I just provide another way to quick fix that might be simpler: 
replace [Line 247](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/authorization/role_assignment_resource.go#L247) with
```
id, err := parseRoleAssignmentId(d.Id())
```
then we could construct the correct resource ID.

**Comment by aristosvo** [comment]
@njuCZ I don't see how your suggestion might fix the bug, but happy to help.

I believe this PR is complete and a forward compatible improvement, not just a fix, but correct me please if I'm wrong.

File: azurerm/internal/services/authorization/parse/role_assignment_test.go [hunk 1]
```diff
@@ -176,6 +196,18 @@ func TestRoleAssignmentID(t *testing.T) {
...
+			Expected: &RoleAssignmentId{
+				SubscriptionID:   "12345678-1234-9876-4563-123456789012",
+				ResourceGroup:    "group1",
+				ResourceProvider: "Microsoft.Storage",
+				ResourceScope:    "storageAccounts/nameStorageAccount",
+				ManagementGroup:  "",
+				Name:             "23456781-2349-8764-5631-234567890121",
+				TenantId:         "34567812-3456-7653-6742-345678901234",
+			},
+		},
```
Comment (njuCZ) [inline azurerm/internal/services/authorization/parse/role_assignment_test.go:210]: we should check whether the `ResourceProvider` and `ResourceScope` equals to the expected value

File: azurerm/internal/services/authorization/parse/role_assignment.go [hunk 2]
```diff
@@ -90,6 +99,17 @@ func RoleAssignmentID(input string) (*RoleAssignmentId, error) {
...
 		roleAssignmentId.ResourceGroup = id.ResourceGroup
+		if id.Provider != "Microsoft.Authorization" && id.Provider != "" {
+			roleAssignmentId.ResourceProvider = id.Provider
+			// logic to save resource scope
+			result := strings.Split(input, "/")
+			for k, v := range result {
+				if v == id.Provider && len(result) >= k+1 {
+					roleAssignmentId.ResourceScope = fmt.Sprintf("%s/%s", result[k+1], result[k+2])
+				}
+			}
```
Comment (njuCZ) [inline azurerm/internal/services/authorization/parse/role_assignment.go:110]: Could it work if the resource scope is complicated? for example: /subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/group1/providers/Microsoft.AppPlatform/Spring/spring1/apps/app1

the expected ResourceScope should be `Spring/spring1/apps/app1`?

**Comment by njuCZ** [comment]
@aristosvo thanks for this PR, I just left some personal thought. Hopes to see it merged soon.

File: azurerm/internal/services/authorization/parse/role_assignment_test.go (hunk 1 above)
Comment (aristosvo) [reply azurerm/internal/services/authorization/parse/role_assignment_test.go:210]: Thanks, missed that indeed! 😄

File: azurerm/internal/services/authorization/parse/role_assignment.go (hunk 2 above)
Comment (aristosvo) [reply azurerm/internal/services/authorization/parse/role_assignment.go]: Fixed, WDYT?

**Comment by sdebruyn** [comment]
Can someone help with some kind of script to repair the mess that 2.62.0 created? 
he PR will fix the issue by itself, but all role assignments are gone from the state files.

Importing them is a PITA, especially since the ID is generated by Azure and not something you can easily piece together. So you'll have to query with the Azure CLI to get the IDs to properly import in Terraform. Something that mimicks an "auto-import" feature would be a perfect solution.

I have to fix this in tens of projects and not looking forward to a manual process.

**Comment by aristosvo** [comment]
@sdebruyn Dependent on your experience with the Azure provider forTerraform.., one dangerous and wild option is to remove the logic which is preventing recreating unimported resources for this specific resource, build your custom Azure TF provider and run it against your projects...?

**Comment by aristosvo** [comment]
@sdebruyn I can help you, but please do it under one of 8 open issues regarding this fix, not under the PR!!

**Comment by etaham** [comment]
Can this be patched? The bug causing state corruption for existing role assignments. The existing role assignments are being removed when state is applied. We needed to rollback to 2.61 and then import the assignments to get them back into state.

**Review by katbyte — APPROVED** [review]
Thanks a ton @aristosvo - this LGTM 🚀

**Comment by etaham** [comment]
Thank you!!

**Comment by Richard-CDS** [comment]
> I have a script ready and I'll share it when it succeeds. But I can't pin on .61 anymore since it won't decode azurerm_api_management_subscription from the state anymore with that version (another change in .62)

Just wondering if you were able to share your script to fix the state file?

**Pull request merged**